src/object/srv_obj.c

```diff
@@ -1464,82 +1416,152 @@ ds_obj_punch_local_hdlr(struct obj_punch_in *opi, crt_opcode_t opc,
 	return rc;
 }
 
+/* Handle the punch requests on non-leader */
 void
-ds_obj_punch_handler(crt_rpc_t *rpc)
+ds_obj_tgt_punch_handler(crt_rpc_t *rpc)
 {
-	struct obj_req_disp_arg		*obj_arg = NULL;
 	struct ds_cont_hdl		*cont_hdl = NULL;
 	struct ds_cont_child		*cont = NULL;
 	struct dtx_handle		*dth = NULL;
-	struct dtx_id			*dti_cos = NULL;
 	struct dtx_conflict_entry	 conflict = { 0 };
 	struct obj_punch_in		*opi;
 	uint32_t			 map_version = 0;
-	uint32_t			 flags = 0;
-	int				 tag;
-	int				 dti_cos_count = 0;
-	bool				 dispatch;
-	bool				 resend = false;
 	int				 rc;
 
 	opi = crt_req_get(rpc);
 	D_ASSERT(opi != NULL);
-	dispatch = opi->opi_shard_tgts.ca_arrays != NULL;
+	rc = ds_pre_check(opi->opi_oid, opi->opi_map_ver, opi->opi_co_hdl,
+			  opi->opi_co_uuid, opc_get(rpc->cr_opc),
+			  opi->opi_flags, &opi->opi_dti, &cont_hdl, &cont);
+	if (rc)
+		goto out;
 
-	tag = dss_get_module_info()->dmi_tgt_id;
+	/* Handle resend. */
+	if (opi->opi_flags & ORF_RESEND) {
+		rc = dtx_handle_resend(cont->sc_hdl, &opi->opi_oid,
+				       &opi->opi_dti, opi->opi_dkey_hash, true);
+		if (rc == -DER_ALREADY)
+			D_GOTO(out, rc = 0);
+	}
 
-	rc = ds_check_container(opi->opi_co_hdl, opi->opi_co_uuid,
-				&cont_hdl, &cont);
-	if (rc)
+	/* Start the local transaction */
+	rc = dtx_begin(&opi->opi_dti, &opi->opi_oid, cont->sc_hdl,
+		       opi->opi_epoch, opi->opi_dkey_hash,
+		       &conflict, opi->opi_dti_cos.ca_arrays,
+		       opi->opi_dti_cos.ca_count, opi->opi_map_ver,
+		       DAOS_INTENT_PUNCH, false, &dth);
+	if (rc != 0) {
+		D_ERROR(DF_UOID": Failed to start DTX for punch %d.\n",
+			DP_UOID(opi->opi_oid), rc);
 		D_GOTO(out, rc);
+	}
 
-	if (!(cont_hdl->sch_capas & DAOS_COO_RW))
-		D_GOTO(out, rc = -DER_NO_PERM);
+	/* local RPC handler */
+	rc = obj_local_punch(opi, opc_get(rpc->cr_opc), cont_hdl, cont, dth);
+	if (rc != 0) {
+		D_ERROR(DF_UOID": obj_local_punch failed %d.\n",
+			DP_UOID(opi->opi_oid), rc);
+		D_GOTO(out, rc);
+	}
 
-	D_ASSERT(cont_hdl->sch_pool != NULL);
-	map_version = cont_hdl->sch_pool->spc_map_version;
+	/**
+	 * XXX layer violation, it should be moved to DTX module, will
+	 * handle in following patches.
+	 */
+	if (opi->opi_dti_cos.ca_count > 0) {
+		/* XXX: For non-leader replica, even if we fail to
+		 *	make related modification for some reason,
+		 *	we still need to commit the DTXs for CoS.
+		 *	Because other replica may have already
+		 *	committed them. For leader case, it is
+		 *	not important even if we miss to commit
+		 *	the CoS DTXs, because they are still in
+		 *	CoS cache, and can be committed next time.
+		 */
+		rc = vos_dtx_commit(cont->sc_hdl,
```

Comment:
similar as update case.

Reply:
yes, it is in tgt_punch_handler()

Reply:
"Before we get better solution for resend, I would suggest to keep the resend handling."
Well we already agree dtx is not good place for resend, since undo log overhead. since no one is using it at all. Can I deal with it in the following patch. I am still thinking. sorry.

Reply:
We agree that the DTX is not good solution for resend, but we also agree that we do not have better solution up to now. Currently, we need the resend handling mechanism, if without that, the resent modification will cause DTX trouble.